### DISCUSSION TITLE
How to webtorrent

### DISCUSSION THREAD

**Discussion by SmartFlamz** [community/community#133126] [Other Feature Feedback, Questions, & Ideas]
### Help with Webtorrent

Question

### Body

I'll like to set-up a JavaScript code and utilitize webtorrent to search magnet links for .mkv files and serve it to a specific video element on a webpage. I have no idea on how to do this so please any and every help will be appreciated.

**Comment by luismede** [comment]
I've never tested it, but I believe maybe this library can help you:
https://www.npmjs.com/package/torrent-files-library
on github: https://github.com/jy95/torrent-files-library#readme

Ps: using just webtorrent, I didn't find anything

**Reply by SmartFlamz** [reply]
Thank you so much,  I'll try it now and relay the outcome, please anymore will be much appreciated 🙏🏽

**Reply by luismede** [reply]
Okay, I think this might help you! send updates later. 👊

**Reply by SmartFlamz** [reply]
> I've never tested it, but I believe maybe this library can help you: https://www.npmjs.com/package/torrent-files-library on github: https://github.com/jy95/torrent-files-library#readme
> 
> Ps: using just webtorrent, I didn't find anything



The above library didn't work, it gave errors in the console I think the "torrent-files-library" is no longer maintained. Please if needed I'll share console errors soon, I forgot to save them earlier when trying the method so I'll have to redo the whole process, I'm still new and unexperienced to all these 😅.

**Reply by luismede** [reply]
Damn 😕, it's really old. I believe it is no longer maintained! But unfortunately I didn't find anything to filter .mkv torrents.

In Python I know how to do this, but then I would deviate from your initial idea, which is to use js with webtorrent.
As for being inexperienced. You're on the right path, keep studying and testing your projects! 👏

**Reply by SmartFlamz** [reply]
Thank you very much 😊

**Reply by luismede** [reply]
Hi, @SmartFlamz  How is your project going? 👋

**Reply by SmartFlamz** [reply]
Hi 👋🏽 @luismede thanks for the check-up, I abandoned that project as I couldn't find any method to fully stream torrents. Thanks again.

**Comment by BONY-SL** [comment]
1. Setting Up Your Environment
You'll need to include the WebTorrent library in your project. You can do this by including the CDN link in your HTML file.

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>WebTorrent Streaming</title>
</head>
<body>
    
    <video id="videoPlayer" controls></video>

    
    <script src="https://cdn.jsdelivr.net/npm/webtorrent/webtorrent.min.js"></script>

    
    <script>
        // Your JavaScript code will go here
    </script>
</body>
</html>


2. Writing the JavaScript Code
Now, let's write the JavaScript code to search for magnet links, load the torrent, and serve it to the video element.

document.addEventListener('DOMContentLoaded', function () {
    const client = new WebTorrent();
    const videoElement = document.getElementById('videoPlayer');

    // Replace with your magnet link
    const magnetURI = 'magnet:?xt=urn:btih:YOUR_MAGNET_LINK_HERE';

    // Load the torrent
    client.add(magnetURI, function (torrent) {
        // Find the .mkv file in the torrent files
        const file = torrent.files.find(file => file.name.endsWith('.mkv'));

        if (file) {
            // Stream the file to the video element
            file.renderTo(videoElement);
        } else {
            console.error('No .mkv file found in the torrent.');
        }
    });

    client.on('error', function (err) {
        console.error('WebTorrent error:', err);
    });
});


3. Explanation
WebTorrent Client: We create a new WebTorrent client instance to interact with the torrent network.
Magnet URI: Replace 'YOUR_MAGNET_LINK_HERE' with your magnet link. This is a unique identifier for the torrent you want to load.
Finding .mkv Files: The code searches for a file within the torrent that ends with .mkv.
Rendering to Video Element: If the .mkv file is found, it is rendered directly to the HTML <video> element.

4. Running the Code
Ensure that you replace the placeholder magnet URI with an actual magnet link. You can get magnet links from various torrent sites.
Open the HTML file in your browser, and if the magnet link is valid and the file is found, the video should start streaming in the video player.

5. Considerations
WebRTC Support: This code will only work in modern browsers that support WebRTC.
CORS Issues: Depending on the server and torrent, you might run into CORS issues when trying to stream the video directly.
Performance: Streaming large files might take some time depending on the health of the torrent (number of seeders, etc.).

6. Optional: Adding Magnet Link Input
If you want to allow users to input their own magnet links, you can add an input box and modify the JavaScript to handle dynamic magnet links.

<body>
   <input type="text" id="magnetInput" placeholder="Enter Magnet Link">
   <button id="loadTorrent">Load Torrent</button>
   <video id="videoPlayer" controls></video>

   <script src="https://cdn.jsdelivr.net/npm/webtorrent/webtorrent.min.js"></script>
   <script>
       document.getElementById('loadTorrent').addEventListener('click', function () {
           const client = new WebTorrent();
           const videoElement = document.getElementById('videoPlayer');
           const magnetURI = document.getElementById('magnetInput').value;

           client.add(magnetURI, function (torrent) {
               const file = torrent.files.find(file => file.name.endsWith('.mkv'));

               if (file) {
                   file.renderTo(videoElement);
               } else {
                   console.error('No .mkv file found in the torrent.');
               }
           });

           client.on('error', function (err) {
               console.error('WebTorrent error:', err);
           });
       });
   </script>
</body>
